### PR TITLE
Add GPU memory usage tracker

### PR DESCRIPTION
**Pull request by wenkaidu**

### CODE REVIEW

File: src/include/alloc.h [hunk 1]
```diff
@@ -37,13 +37,28 @@ static ncclResult_t ncclCalloc(T** ptr, size_t nelem) {
   return ncclSuccess;
 }
 
+struct __attribute__ ((aligned(64))) allocationTracker {
+  uint64_t totalAlloc;
```
Comment (gilbertlee-amd) [inline src/include/alloc.h:41]: Can we just use a union with one field being 64 chars?

File: src/include/alloc.h [hunk 2]
```diff
@@ -37,13 +37,28 @@ static ncclResult_t ncclCalloc(T** ptr, size_t nelem) {
...
 static ncclResult_t ncclCudaCalloc(T** ptr, size_t nelem, bool isFineGrain = false) {
   if (isFineGrain)
     CUDACHECK(hipExtMallocWithFlags((void**)ptr, nelem*sizeof(T), hipDeviceMallocFinegrained));
   else
     CUDACHECK(hipMalloc(ptr, nelem*sizeof(T)));
   CUDACHECK(hipMemset(*ptr, 0, nelem*sizeof(T)));
+  int dev;
+  CUDACHECK(hipGetDevice(&dev));
+  if (dev < MAX_ALLOC_TRACK_NGPU) {
+    __atomic_fetch_add(&allocTracker[dev].totalAlloc, 1, __ATOMIC_SEQ_CST);
```
Comment (gilbertlee-amd) [inline src/include/alloc.h:59]: The number of allocs is being counted, but I don't think it's being reported anywhere.

File: src/init.cc [hunk 3]
```diff
@@ -58,6 +58,8 @@ NCCL_PARAM(CheckPointers, "CHECK_POINTERS", 0);
 ncclNet_t* ncclNet = NULL;
 ncclCollNet_t* ncclCollNet = NULL;
 
+struct allocationTracker allocTracker[MAX_ALLOC_TRACK_NGPU];
```
Comment (gilbertlee-amd) [inline src/init.cc:61]: Might need to initialize this with = {};

File: src/include/alloc.h [hunk 4]
```diff
@@ -37,13 +37,28 @@ static ncclResult_t ncclCalloc(T** ptr, size_t nelem) {
...
+    __atomic_fetch_add(&allocTracker[dev].totalAllocSize, nelem*sizeof(T), __ATOMIC_SEQ_CST);
```
Comment (gilbertlee-amd) [inline src/include/alloc.h:60]: These values aren't reset when comm gets destroyed - is that intentional?

File: src/include/alloc.h (hunk 1 above)
Comment (wenkaidu) [reply src/include/alloc.h]: done

File: src/include/alloc.h (hunk 2 above)
Comment (wenkaidu) [reply src/include/alloc.h:63]: The information is not useful outside RCCL. Thus I decided not to print it.

File: src/include/alloc.h (hunk 4 above)
Comment (wenkaidu) [reply src/include/alloc.h:64]: I missed it. I added reset to init, as sometimes framework do not call comm destroy

File: src/init.cc (hunk 3 above)
Comment (wenkaidu) [reply src/init.cc]: done

File: src/include/alloc.h (hunk 2 above)
Comment (gilbertlee-amd) [reply src/include/alloc.h:63]: But why even track then?  How do you check that value?

**Review by gilbertlee-amd — APPROVED** [review]

File: src/include/alloc.h (hunk 2 above)
Comment (wenkaidu) [reply src/include/alloc.h:63]: We can check with gdb. Anyway, original thought was to use counter to track hipMalloc/hipFree pair so we can detect leak. But this requires create a wrapper and change all location of hipFree. We can discuss offline to see if this is worth doing.

**Pull request merged**